File: languages/go/client/availability_test.go
//go:build e2e

package client

import (
	"testing"

	"github.com/bloock/bloock-sdk-go/v2/entity/availability"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestAvailability(t *testing.T) {
	InitDevSdk()

	t.Run("publish hosted", func(t *testing.T) {
		payload := "Hello world"
		recordClient := NewRecordClient()
		record, err := recordClient.FromString(payload).Build()
		require.NoError(t, err)

		availabilityClient := NewAvailabilityClient()
		result, err := availabilityClient.Publish(record, availability.NewHostedPublisher())
		require.NoError(t, err)
		assert.NotEmpty(t, result.ID)
	})

	t.Run("retrieve hosted", func(t *testing.T) {
		payload := "Hello world"
		recordClient := NewRecordClient()
		record, err := recordClient.FromString(payload).Build()
		require.NoError(t, err)

		recordHash, err := record.GetHash()
		require.NoError(t, err)

		availabilityClient := NewAvailabilityClient()
		res, err := availabilityClient.Publish(record, availability.NewHostedPublisher())
		require.NoError(t, err)

		result, err := availabilityClient.Retrieve(availability.NewHostedLoader(res.ID))
		require.NoError(t, err)

		resultHash, err := result.GetHash()
		require.NoError(t, err)
		assert.Equal(t, recordHash, resultHash)
	})

	t.Run("publish ipfs", func(t *testing.T) {
		payload := "Hello world"
		recordClient := NewRecordClient()
		record, err := recordClient.FromString(payload).Build()
		require.NoError(t, err)

		availabilityClient := NewAvailabilityClient()
		result, err := availabilityClient.Publish(record, availability.NewIpfsPublisher())
		require.NoError(t, err)
		assert.NotEmpty(t, result.ID)
	})

	t.Run("retrieve ipfs", func(t *testing.T) {
		payload := "Hello world"
		recordClient := NewRecordClient()
		record, err := recordClient.FromString(payload).Build()
		require.NoError(t, err)

		recordHash, err := record.GetHash()
		require.NoError(t, err)

		availabilityClient := NewAvailabilityClient()
		res, err := availabilityClient.Publish(record, availability.NewIpfsPublisher())
		require.NoError(t, err)

		result, err := availabilityClient.Retrieve(availability.NewIpfsLoader(res.ID))
		require.NoError(t, err)

		resultHash, err := result.GetHash()
		require.NoError(t, err)
		assert.Equal(t, recordHash, resultHash)
	})

	t.Run("publish and retrieve from ipns", func(t *testing.T) {
		payload := "Hello world"
		recordClient := NewRecordClient()
		record, err := recordClient.FromString(payload).Build()
		require.NoError(t, err)

		recordHash, err := record.GetHash()
		require.NoError(t, err)

		availabilityClient := NewAvailabilityClient()
		response, err := availabilityClient.Publish(record, availability.NewIpnsPublisher())
		require.NoError(t, err)
		assert.NotEmpty(t, response.ID)
		assert.NotEmpty(t, response.IpnsKey.KeyID)

		result, err := availabilityClient.Retrieve(availability.NewIpnsLoader(response.ID))
		require.NoError(t, err)

		resultHash, err := result.GetHash()
		require.NoError(t, err)
		assert.Equal(t, recordHash, resultHash)

		payloadUpdate := "Bye Bye"
		recordUpdate, err := recordClient.FromString(payloadUpdate).Build()
		require.NoError(t, err)

		responseUpdated, err := availabilityClient.Publish(recordUpdate, availability.UpdateIpnsPublisher(response.IpnsKey))
		assert.NoError(t, err)
		assert.Equal(t, response.ID, responseUpdated.ID)
		assert.Equal(t, response.IpnsKey, responseUpdated.IpnsKey)
	})
}

File: languages/js/test/e2e/availability.test.ts
import { describe, test, expect } from "@jest/globals";
import { initDevSdk } from "./util";
import {
  AvailabilityClient,
  HostedLoader,
  HostedPublisher,
  IpfsLoader,
  IpfsPublisher,
  RecordClient,
  IpnsPublisher,
  IpnsKey,
  IpnsLoader,
} from "../../dist";

describe("Availability Tests", () => {
  test("publish hosted", async () => {
    initDevSdk();

    let payload = "Hello world";
    let recordClient = new RecordClient();
    let record = await recordClient.fromString(payload).build();

    let availabilityClient = new AvailabilityClient();
    let result = await availabilityClient.publish(
      record,
      new HostedPublisher()
    );

    expect(result.id).toBeTruthy();
    expect(result.ipnsKey).toBeUndefined();
  });

  test("retrieve hosted", async () => {
    initDevSdk();

    let payload = "Hello world";
    let recordClient = new RecordClient();
    let record = await recordClient.fromString(payload).build();

    let hash = await record.getHash();

    let availabilityClient = new AvailabilityClient();
    let res = await availabilityClient.publish(record, new HostedPublisher());

    let result = await availabilityClient.retrieve(new HostedLoader(res.id));
    let resultHash = await result.getHash();

    expect(resultHash).toBe(hash);
  });

  test("publish ipfs", async () => {
    initDevSdk();

    let payload = "Hello world";
    let recordClient = new RecordClient();
    let record = await recordClient.fromString(payload).build();

    let availabilityClient = new AvailabilityClient();
    let result = await availabilityClient.publish(record, new IpfsPublisher());

    expect(result.id).toBeTruthy();
    expect(result.ipnsKey).toBeUndefined();
  });

  test("retrieve ipfs", async () => {
    initDevSdk();

    let payload = "Hello world";
    let recordClient = new RecordClient();
    let record = await recordClient.fromString(payload).build();

    let hash = await record.getHash();

    let availabilityClient = new AvailabilityClient();
    let res = await availabilityClient.publish(record, new IpfsPublisher());

    let result = await availabilityClient.retrieve(new IpfsLoader(res.id));
    let resultHash = await result.getHash();

    expect(resultHash).toBe(hash);
  });

  test("publish ipns", async () => {
    initDevSdk();

    let payload = "Hello world";
    let recordClient = new RecordClient();
    let record = await recordClient.fromString(payload).build();
    let recordHash = await record.getHash()

    let availabilityClient = new AvailabilityClient();
    let result = await availabilityClient.publish(record, new IpnsPublisher());

    expect(result.id).toBeTruthy();
    expect(result.ipnsKey).toBeTruthy();

    let resultRetrieved = await availabilityClient.retrieve(new IpnsLoader(result.id));
    let resultHash = await resultRetrieved.getHash();

    expect(resultHash).toBe(recordHash);

    let recordUpdated = await recordClient.fromString("Bye Bye").build();
    if (result.ipnsKey) {
      let resultUpdated = await availabilityClient.publish(recordUpdated, new IpnsPublisher(new IpnsKey(result.ipnsKey.keyID)));

      expect(resultUpdated.id).toBeTruthy();
      expect(resultUpdated.ipnsKey).toBeTruthy();
    }
  });
});
